scipy/interpolate/_aaa.py

```diff
@@ -0,0 +1,353 @@
+    >>> sample_points = np.linspace(-1.5, 1.5, num=100)
+    >>> r = AAA(gamma, sample_points)
+    >>> z = np.linspace(-3.5, 4.5, num=1000)
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(z, gamma(z), label="Gamma")
+    >>> ax.plot(sample_points, gamma(sample_points), label="Sample points")
+    >>> ax.plot(z, r(z).real, '--', label="AAA approximation")
+    >>> ax.set(xlabel="z", ylabel="r(z)", ylim=[-8, 8], xlim=[-3.5, 4.5])
+    >>> ax.legend()
+    >>> plt.show()
+
+    We can also view the poles of the rational approximation and their residue:
+
+    >>> for i in np.argsort(r.poles):
+    ...     print(f"{r.poles[i]=:.3e}, {r.residues[i]=:.3e}")
+    r.poles[i]=-3.816e+00+0.000e+00j, r.residues[i]=3.658e-02+0.000e+00j
+    r.poles[i]=-3.003e+00+0.000e+00j, r.residues[i]=-1.692e-01-0.000e+00j
+    r.poles[i]=-2.000e+00+0.000e+00j, r.residues[i]=5.000e-01+0.000e+00j
+    r.poles[i]=-1.000e+00+0.000e+00j, r.residues[i]=-1.000e+00+0.000e+00j
+    r.poles[i]=5.858e-17+0.000e+00j, r.residues[i]=1.000e+00+0.000e+00j  # may vary
+    r.poles[i]=4.775e+00-3.069e+00j, r.residues[i]=-8.113e-01-2.302e+00j
+    r.poles[i]=4.775e+00+3.069e+00j, r.residues[i]=-8.113e-01+2.302e+00j
+    r.poles[i]=5.291e+00-9.737e-01j, r.residues[i]=8.733e-01+1.070e+01j
+    r.poles[i]=5.291e+00+9.737e-01j, r.residues[i]=8.733e-01-1.070e+01j
+
+    For the second example, we call `AAA` with a spiral of 1000 points wind 7.5 times
+    around the origin in the complex plane.
+
+    >>> z = np.exp(np.linspace(-0.5, 0.5 + 15j*np.pi, 1000))
+    >>> r = AAA(lambda z: np.tan(np.pi*z/2), z)
+
+    We see that AAA takes 12 steps to converge with the following errors:
+
+    >>> r.errors.size
+    12
+    >>> r.errors
+    array([2.49261500e+01+0.j, 4.28045609e+01+0.j, 1.71346935e+01+0.j,
+        8.65055336e-02+0.j, 1.27106444e-02+0.j, 9.90889874e-04+0.j,
+        5.86910543e-05+0.j, 1.28735561e-06+0.j, 3.57007532e-08+0.j,
+        6.37019486e-10+0.j, 1.67119305e-11+0.j, 1.17128023e-13+0.j])
+
+    We can also plot the computed poles:
+
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(z.real, z.imag, '.', markersize=2, label="Sample points")
+    >>> ax.plot(r.poles.real, r.poles.imag, '.', markersize=5, label="Computed poles")
+    >>> ax.set(xlim=[-3.5, 3.5], ylim=[-3.5, 3.5], aspect="equal")
+    >>> ax.legend()
+    >>> plt.show()
+    """
+    def __init__(self, f, z, *, rtol=1e-13, max_terms=100):
+        # input validation
+        z = np.asarray(z).ravel()
+        if callable(f):
+            f = f(z)
+        else:
+            f = np.asarray(f).ravel()
+
+            if f.size != z.size:
+                raise ValueError("`f` and `z` must be the same size.")
+
+        # Remove infinite or NaN function values and repeated entries
+        to_keep = (np.isfinite(f)) & (~np.isnan(f))
+        f = f[to_keep]
+        z = z[to_keep]
+        z, uni = np.unique(z, return_index=True)
+        f = f[uni]
+
+        # Initialization for AAA iteration
```

Comment:
Consider moving the majority of the computations from `__init__` to a helper function

Reply:
Done